### PR TITLE
Local exceptions for PaletteManipulator

### PR DESCRIPTION
**Pull request by aschempp**
I've run into the problem that DcaManipulator throws an `InvalidArgumentException`, which is catched by the `DcaLoader` because `FileLocator` throws the same.

1. we should always use local exceptions
2. Added logging if an exception is catched

### CODE REVIEW

**Comment by aschempp** [comment]
@leofeyer ping

**Comment by leofeyer** [comment]
I guess you will have to re-create the PR against an infinite branch. I cannot reopen it. :cry:

**Comment by aschempp** [comment]
Or you just merge it into 4.2.2? ;-)

**Comment by leofeyer** [comment]
I can't. It's a feature.

**Comment by aschempp** [comment]
well then merge it to develop…

**Comment by leofeyer** [comment]
Ok, will do.

File: src/Resources/contao/library/Contao/DcaLoader.php [hunk 1]
```diff
@@ -85,7 +85,10 @@ public function load($blnNoCache=false)
 					include $file;
 				}
 			}
-			catch (\InvalidArgumentException $e) {}
+			catch (\InvalidArgumentException $e)
+			{
+				\System::getContainer()->get('monolog.logger.contao')->notice($e->getMessage(), ['exception' => $e]);
```
Comment (leofeyer) [inline src/Resources/contao/library/Contao/DcaLoader.php]: Why did you add this here? It does not seem to be related to the PR.

Comment (aschempp) [inline src/Resources/contao/library/Contao/DcaLoader.php]: Yes it is, that's point 2 in my comment

Comment (leofeyer) [inline src/Resources/contao/library/Contao/DcaLoader.php]: But this will also log exceptions for every bundle or module that does not have a `dca/` folder! What exactly do you want to log?

Comment (aschempp) [inline src/Resources/contao/library/Contao/DcaLoader.php]: It would log if no such file was found in any `dca` folder, right? Which should theoretically not happen because we would not find the table…

My issue was that an `InvalidArgumentException` was thrown by a DCA file (in this case `PaletteManipulator` but it could be anything), and there was no way to figure out what went wrong and why my file was not loaded.

Maybe we should create a PR for Symfony to throw a `ResourceNotFoundException` in the next version?

Comment (leofeyer) [inline src/Resources/contao/library/Contao/DcaLoader.php]: > Maybe we should create a PR for Symfony to throw a `ResourceNotFoundException` in the next version?

The Symfony `FileLocator` class throws the `\InvalidArgumentException`, therefore we cannot change this (except by replacing the class). But the following is wrong anyway:

```php
class PaletteNotFoundException extends \InvalidArgumentException
```

All palette errors are `\LogicException`s and if the palette exceptions extended these, they would not be caught by the file locator routine.

Comment (aschempp) [inline src/Resources/contao/library/Contao/DcaLoader.php]: Well we can't change that for BC reasons, extending from `\InvalidArgumentException` is BC.

Comment (aschempp) [inline src/Resources/contao/library/Contao/DcaLoader.php]: And I think specifying an invalid palette is an `InvalidArgumentException` ;-)

Comment (leofeyer) [inline src/Resources/contao/library/Contao/DcaLoader.php]: Both wrong IMHO.

> Well we can't change that for BC reasons, extending from `\InvalidArgumentException` is BC.

The `PaletteNotFoundException` class is only added in this PR, therefore there cannot be a BC break.

> And I think specifying an invalid palette is an `InvalidArgumentException` ;-)

Not necessarily. An invalid argument is something that is passed at runtime. A configuration error in a file (in this case a DCA file) is a logic error.

Comment (aschempp) [inline src/Resources/contao/library/Contao/DcaLoader.php]: > The PaletteNotFoundException class is only added in this PR, therefore there cannot be a BC break.

But previously it threw an `InvalidArgumentException`, so we must stay compatible with that if someone already catched it.

Comment (leofeyer) [inline src/Resources/contao/library/Contao/DcaLoader.php]: I see.
